core/lib/web3_decl/src/client/mod.rs

```diff
@@ -0,0 +1,357 @@
+//! L2 HTTP client.
+
+use std::{
+    collections::HashSet,
+    fmt,
+    num::NonZeroUsize,
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+
+use async_trait::async_trait;
+use jsonrpsee::{
+    core::{
+        client::{BatchResponse, ClientT, Error, Subscription, SubscriptionClientT},
+        params::BatchRequestBuilder,
+        traits::ToRpcParams,
+    },
+    http_client::{HttpClient, HttpClientBuilder},
+};
+use serde::de::DeserializeOwned;
+use tokio::time::Instant;
+
+use self::metrics::METRICS;
+
+mod metrics;
+#[cfg(test)]
+mod tests;
+
+#[derive(Debug, Clone, Copy)]
+enum CallOrigin<'a> {
+    Notification(&'a str),
+    Request(&'a str),
+    BatchRequest(&'a BatchRequestBuilder<'a>),
+    Subscription(&'a str),
+}
+
+impl<'a> CallOrigin<'a> {
+    fn request_count(self) -> usize {
+        match self {
+            Self::BatchRequest(batch) => batch.iter().count(),
+            _ => 1,
+        }
+    }
+
+    fn distinct_method_names(self) -> HashSet<&'a str> {
+        match self {
+            Self::Notification(name) | Self::Request(name) | Self::Subscription(name) => {
+                HashSet::from([name])
+            }
+            Self::BatchRequest(batch) => batch.iter().map(|(name, _)| name).collect(),
+        }
+    }
+}
+
+impl fmt::Display for CallOrigin<'_> {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Notification(name) => write!(formatter, "notification `{name}`"),
+            Self::Request(name) => write!(formatter, "request `{name}`"),
+            Self::BatchRequest(req) => {
+                let method_names: Vec<_> = req.iter().map(|(name, _)| name).collect();
+                write!(formatter, "batch request {method_names:?}")
+            }
+            Self::Subscription(name) => write!(formatter, "subscription `{name}`"),
+        }
+    }
+}
+
+/// JSON-RPC client for the main node with built-in middleware support.
+///
+/// The client should be used instead of `HttpClient` etc. A single instance of the client should be built
+/// and shared among all tasks run by the node in order to correctly rate-limit requests.
+#[derive(Debug, Clone)]
+pub struct L2Client {
+    inner: HttpClient,
+    rate_limit: SharedRateLimit,
+    component_name: &'static str,
+}
```

Comment:
Maybe stupid question, but still. Looking at jsonrpsee docs, `HttpClientBuilder` [allows setting a middleware](https://docs.rs/jsonrpsee-http-client/0.22.3/jsonrpsee_http_client/struct.HttpClientBuilder.html#method.set_http_middleware), won't it be possible to implement rate limiting through it to keep using `HttpClient` and avoid creating a wrapper client?

Reply:
I've attempted to do this initially (see first commits of this PR), but have run into several issues:

- `HttpClient` allows to set HTTP-level middleware, not an RPC-level one, which I feel is more appropriate for rate limiting.
- There are hypothetical cases like automated "debouncing" batching of requests I've described [above](https://github.com/matter-labs/zksync-era/pull/1500#discussion_r1540952422), which wouldn't work with a HTTP-level middleware w/o significant hacks.
- Observability (logging, metrics) are also easier with RPC-level middleware.
- Naming the HTTP middleware type is problematic if middleware is complex. Tower provides [a boxed cloneable version of services](https://docs.rs/tower/latest/tower/util/struct.BoxCloneService.html), but it doesn't fit (it's unconditionally `!Sync`, and `Sync` is required for `HttpClient<_>` to implement RPC traits).

I'll probably add this explanation as a comment, so that it's not lost in the future.